### PR TITLE
Add convenience module tools

### PR DESCRIPTION
**Pull request by j6k4m8**
This moves our long-standing numpy-style indexing library and URI-mapping to the `intern` library from our various other repos. Experimental work will continue in those repos, but these (mostly) stable APIs can now live in _intern_.

# URI-Mapping

This adds support for the fully-qualified URIs we've been using in metadata to reproducibly distinguish between Remotes:

```python
dvid_data = intern.array("dvid://https://my-dvid-instance.org/82252477...")
boss_data = intern.array(
    "bossdb://https://api.bossdb.io/my_collection/my_experiment/my_channel", 
    token="public"
)
```

# Numpy-style indexing

This style of indexing is growing in popularity and is a great help for those who are new to out-of-RAM large arrays.

```python
>>> boss_data = intern.array(
...    "bossdb://https://api.bossdb.io/my_collection/my_experiment/my_channel"
... )
>>> boss_data.shape
(19587, 244590, 244590)
```

Proxy for `intern.remote.BossRemote#get_cutout`:
```python
>>> cutout = boss_data[512:1024, 512:1024, 512:1024]
```

Proxy for `intern.remote.BossRemote#create_cutout`:
```python
>>> boss_data[512:1024, 512:1024, 512:1024] = my_numpy_data
```

# Create new resources with `create_new`

By passing `create_new=True` and dataset extents, the convenience module will create new Collection, Experiment, CoordinateFrame, and Channel resources:

```python
data = array(
    "bossdb://matelsky2020/tutorial/em", 
    create_new=True,
    extents=[100, 1000, 1000],
    voxel_size=[40, 4, 4], voxel_unit="nanometers", 
)
```

(Optionally specify a `description` and `dtype`.)

### CODE REVIEW

File: intern/convenience/array.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+from intern.utils.parallel import block_compute
+
+# A named tuple that represents a bossDB URI.
+bossdbURI = namedtuple(
+    "bossdbURI", ["collection", "experiment", "channel", "resolution"]
+)
+
+
+class VolumeProvider:
+    def get_channel(self, channel: str, collection: str, experiment: str):
```
Comment (movestill) [inline intern/convenience/array.py:42]: These type annotations make the rest of intern look naked! 😀

File: setup.py [hunk 2]
```diff
@@ -5,57 +5,57 @@
...
+        "microns",
+    ],
+    packages=find_packages(exclude=["docs", "tests*"]),
     include_package_data=True,
-    author='Johns Hopkins University Applied Physics Laboratory',
+    author="Johns Hopkins University Applied Physics Laboratory",
     install_requires=install_requires,
     dependency_links=dependency_links,
-    author_email='iarpamicrons@jhuapl.edu'
+    author_email="iarpamicrons@jhuapl.edu",
```
[the remark this answers is not in the document]
Comment (j6k4m8) [reply setup.py:60]: Good question. Not sure what we have set up, but it makes sense for this to point bossward

File: intern/convenience/array.py (hunk 1 above)
Comment (j6k4m8) [reply intern/convenience/array.py]: heh, type-anno'ing the rest of the library is gonna be a long journey...

**Comment by j6k4m8** [comment]
Ah, just realized tests are failing because I don't have the same mock infrastructure as the intern config-file mocks. working on it...

**Comment by j6k4m8** [comment]
This latest update may be a bit controversial, as it's by far the most opinionated code in the repo. In order to minimize the startup time to using `intern` and BossDB, the convenience module makes a few assumptions already. This latest commit adds default BossRemote options. In other words, if you DON'T have a `~/.intern/intern.cfg` file and you don't specify BossRemote options to the array constructor, it will _assume_ that you want an anonymous (public-user) Remote that points at the BossDB.io API.

I think this is fairly safe as this is \*the\* BossDB as far as the everyday user is concerned, but it's worth consideration here.

**Pull request merged**